src/libs/actions/IOU.js

```diff
@@ -131,10 +132,15 @@ function createIOUSplit(params) {
  * @param {Number} params.chatReportID
  * @param {String} params.transactionID
  * @param {String} params.comment
+ * @param {Object} params.iou
  */
 function rejectTransaction({
-    reportID, chatReportID, transactionID, comment,
+    reportID, chatReportID, transactionID, comment, iou,
 }) {
```

Comment:
I think an improvement here would be to not send the entire IOU state object. If we only need the `transactionsBeingRejected` value then lets just make that the param.

Reply:
Actually, we need the entire object for removing the transaction that's completed rejection in the finally block.
 
I need an Onyx feature 'set one but don't affect others' which doesn't exist.

So a workaround is to pass the entire IOU object and use the merge method.
[more details here](https://github.com/Expensify/Expensify.cash/pull/3498/files/2aa493b144cf298c2dc63d45aae0aeef41a38a2b#diff-daa865a27f5c734becece9770029064b7cd62b56034d54cb2eb0279fa6cc548c)

Reply:
Hmm, are you looking for Onyx.merge? That will update keys, but won't remove others.

Reply:
No, I'll explain with an example.
Assume this is the state of IOU object.
```
IOU : {
   transactionsBeingRejected : ['1000', '1001'],
   loading: false,
   ...
}
```
## Onyx.merge
Let's say the transaction id 1000 is sucessfully rejected, then I need to remove it from the array.
```
Onyx.merge(IOU, {
    transactionsBeingRejected: withoutID1000
})
```
Merge will not remove id 1000 from the object.

## Onyx.set
```
Onyx.set(IOU, {
   transactionsBeingRejected: withoutID1000
})
```
Set will remove id 1000, but other keys of IOU like loading etc are now lost.
## This is what worked
```
Onyx.set(IOU, {
    ...iou,
    transactionsBeingRejected: withoutID1000
});
```
Hence I need the entire IOU object.

Reply:
Okay ignore that ^, I can see you using merge below :)

Alex and I had a quick chat and realized that array probably isn't the best data structure here, for the reason that you mentioned (array values [don't update as we desire](https://github.com/Expensify/react-native-onyx/blob/1e82e592032c6d0ede8e40f08beb6be790d149e8/lib/Onyx.js#L575-L576)). Instead, we should store the rejectingIDs as an object, like we do almost everywhere else within the app.

This way, the rejectingIDs object can be updated one by one, and we don't need to group them all manually. It looks like we just need to merge an[ ID with a value of null ](https://github.com/Expensify/react-native-onyx/blob/1e82e592032c6d0ede8e40f08beb6be790d149e8/lib/Onyx.js#L583-L592)for it to be cleared. Please let me know if that doesn't make sense!

Reply:
Alright, so this should work?
```
iou: PropTypes.shape({
    transactionsBeingRejected: PropTypes.shape({
        id: PropTypes.string,
    }),
}),
```

Reply:
_I believe so_. I think we complicated things by suggesting array in the first place. But please let us know if you run into any other issues!

Reply:
My bad, 
```
iou: PropTypes.shape({
    transactionsBeingRejected: PropTypes.shape({
        id: PropTypes.string,
    }),
}),
```
is equivalent of 
```
iou: PropTypes.shape({
    transactionsBeingRejected: PropTypes.string,
}),
```
**As both can store one ID at a given time only.**
I'll need to use an array of objects then.
But since merge on array is append only, we're back to square one.

Reply:
If a user rejects multiple IOUs at same time, then we need to store multiple transaction IDs (since they're all loading),
and arrays is the only solution I can think of for javascript.

Reply:
I'm not following the reason for this structure not being possible:

// 123 being rejected
```
iou: {
    loading: false,
    creatingIOUTransaction: false,
    transactionsBeingRejected {
        123: true
    }
}
```

// 123 & 456 being rejected
```
iou: {
    loading: false,
    creatingIOUTransaction: false,
    transactionsBeingRejected {
        123: true,
        456: true,
    }
}
```

// 123 is no longer in progress, so we delete by setting null
```
iou: {
    loading: false,
    creatingIOUTransaction: false,
    transactionsBeingRejected {
        123: null,
        456: true,
    }
}
```

Reply:
Makes sense. 
When do the null values get removed? For me, they don't ever seem to get deleted.

![image](https://user-images.githubusercontent.com/29673073/121924471-98cd6580-cd44-11eb-8a17-b3b14b4ef9c9.png)

Reply:
Hmm, I assumed they would be based [upon this comment ](https://github.com/Expensify/react-native-onyx/blob/1e82e592032c6d0ede8e40f08beb6be790d149e8/lib/Onyx.js#L588-L590)- -but it looks like that only applies to top-level keys?  🤔

Reply:
It might be worth posting in the open-source slack channel to get a second opinion here.

Reply:
yeah, with my observations only top-level keys are being deleted.

Reply:
Apparently, null keys aren't removed recursively.
So they don't get deleted.

I see these as possible options now
## Array implementation (current).
Cons: Passing entire IOU object.
## Add the transactionIDs directly to IOU key in the top level.
Cons: Possibly pollutes IOU key state.
## Create a new onyx key
Add transactionIDs in top level as objects or array.
## Modify Onxy to recursively delete null keys.
Need to spend time :)

Reply:
Let's create a new Onyx key. The biggest concern was that we were passing the `iou` Onyx state through a function, so as long as we avoid that it seems fine.

Reply:
Once nested keys with null are deleted we can add it back to `iou` if we really need to.

Reply:
I'll store the IDs in an array. 
Setting them as null to be delete might not stay in the future, especially when a delete function is created. [Related issue](https://github.com/Expensify/Expensify.cash/issues/3490)

I hope this is fine :)

Reply:
> I'll store the IDs in an array. Setting them as null to delete might not stay in the future. Especially if a delete function is created

I still think we should store it as a new key, as discussed previously. We should not pre-optimize for a decision that isn't yet confirmed and may not be implemented.

Also, the array solution does not solve the concern that local storage logic is being handled outside of Onyx.

Reply:
Alright, makes sense.
So a new key with this structure
```
transactionsBeingRejected : {
    123: true,
    456: true,
}
```

and set an id as null for deleting.

Reply:
@Julesssss 
Please review the changes. I think we're good to merge now.